kanidmd/src/lib/idm/oauth2.rs

```diff
@@ -0,0 +1,1035 @@
+                })
+            })?;
+
+        // Get the first :, it should be our delim.
+        //
+        let mut split_iter = authz.split(':');
+
+        let client_id = split_iter.next().ok_or_else(|| {
+            ladmin_error!(audit, "Basic authz invalid format (corrupt input?)");
+            Oauth2Error::AuthenticationRequired
+        })?;
+        let secret = split_iter.next().ok_or_else(|| {
+            ladmin_error!(audit, "Basic authz invalid format (missing ':' seperator?)");
+            Oauth2Error::AuthenticationRequired
+        })?;
+
+        // Get the o2rs for the handle.
+        let o2rs = self.inner.rs_set.get(client_id).ok_or_else(|| {
+            ladmin_warning!(audit, "Invalid oauth2 client_id");
+            Oauth2Error::AuthenticationRequired
+        })?;
+
+        // check the secret.
+        let o2rs_fernet = match o2rs {
+            Oauth2RS::Basic(rsbasic) => {
+                if rsbasic.authz_secret != secret {
+                    lsecurity!(audit, "Invalid oauth2 client_id secret");
+                    return Err(Oauth2Error::AuthenticationRequired);
+                }
+                // We are authenticated! Yay! Now we can actually check things ...
+                &rsbasic.token_fernet
+            }
+        };
+
+        // Check the token_req is within the valid time, and correctly signed for
+        // this client.
+
+        let code_xchg: TokenExchangeCode = o2rs_fernet
+            .decrypt_at_time(&token_req.code, Some(60), ct.as_secs())
+            .map_err(|_| {
+                ladmin_error!(audit, "Failed to decrypt token exchange request");
+                Oauth2Error::InvalidRequest
+            })
+            .and_then(|data| {
+                serde_json::from_slice(&data).map_err(|e| {
+                    ladmin_error!(audit, "Failed to deserialise token exchange code - {:?}", e);
+                    Oauth2Error::InvalidRequest
+                })
+            })?;
+
+        // Validate the code_verifier
+        let mut hasher = sha::Sha256::new();
+        hasher.update(token_req.code_verifier.as_bytes());
+        let code_verifier_hash: Vec<u8> = hasher.finish().iter().copied().collect();
+
+        if code_xchg.code_challenge.0 != code_verifier_hash {
+            lsecurity!(
+                audit,
+                "PKCE code verification failed - this may indicate malicious activity"
+            );
+            return Err(Oauth2Error::InvalidRequest);
+        }
+
+        // Validate the redirect_uri is the same as the original.
+        if token_req.redirect_uri != code_xchg.redirect_uri {
+            ladmin_warning!(
+                audit,
+                "Invalid oauth2 redirect_uri (differs from original request uri)"
+            );
+            return Err(Oauth2Error::InvalidRequest);
+        }
+
+        // We are now GOOD TO GO!
+        // Use this to grant the access token response.
+        let odt_ct = OffsetDateTime::unix_epoch() + ct;
```

Comment:
current time. Time is always a parameter in our functions because that allows us to control time during all tests so we can simulate time based events.